src/Microsoft.Health.Dicom.SqlServer/Features/Schema/VersionedCache.cs

```diff
@@ -0,0 +1,69 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using Microsoft.Health.Dicom.SqlServer.Exceptions;
+
+namespace Microsoft.Health.Dicom.SqlServer.Features.Schema
+{
+    internal sealed class VersionedCache<T> : IDisposable where T : class, IVersioned
```

Comment:
* Do we want to cache the version? We don't need to handle the schema upgrade scenario now, but when we need to, we couldn't cache it.
* Could we use `Lazy<T>` to simplify the logic to handle with cached entity?

Reply:
I would really like to, so I don't have to think about the locking semantics, but we also need the ability to:
1. Invoke the factory "asynchronously"
2. Cancel the factory invocation.

There are some examples of doing it asynchronously, but cancellation poses a problem because of how you have to capture the cancellation token. So I've kept my double-checked locking with some changes based on the source code for [Lazy\<T\>](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Lazy.cs).